bee-common/bee-packable/src/packable/bounded_int.rs

```diff
@@ -0,0 +1,122 @@
+use crate::{
+    coerce::CoerceInfallible,
+    error::{PackError, UnpackError},
+    packer::Packer,
+    unpacker::Unpacker,
+    Packable,
+};
+
+use core::convert::{Infallible, TryFrom, TryInto};
+
+/// Trait that provides an interface for bounded integers.
+pub trait Bounded {
+    /// The type used to define the bounds.
+    type Bounds;
+
+    /// Minimum bounded value.
+    const MIN: Self::Bounds;
+
+    /// Maximum bounded value.
+    const MAX: Self::Bounds;
+
+    /// Actual value.
+    fn value(&self) -> Self::Bounds;
+}
+
+macro_rules! bounded_int {
+    ($(#[$ty_doc:meta])* $ty:ident, $(#[$err_doc:meta])* $err:ident, $int:ident) => {
+        $(#[$err_doc])*
+        #[derive(Debug, Clone, Copy, PartialEq, Eq)]
+        pub struct $err<const MIN: $int, const MAX: $int>(pub $int);
+
+        impl<const MIN: $int, const MAX: $int> Bounded for $err<MIN, MAX> {
+            type Bounds = $int;
+
+            const MIN: $int = MIN;
+            const MAX: $int = MAX;
+
+            fn value(&self) -> $int {
+                self.0
+            }
+        }
+
+        $(#[$ty_doc])*
+        #[derive(Debug, Clone, Copy, PartialEq, Eq)]
+        pub struct $ty<const MIN: $int, const MAX: $int>($int);
+
+        impl<const MIN: $int, const MAX: $int> Bounded for $ty<MIN, MAX> {
+            type Bounds = $int;
+
+            const MIN: $int = MIN;
+            const MAX: $int = MAX;
+
+            fn value(&self) -> $int {
+                self.0
+            }
+        }
+
+        impl<const MIN: $int, const MAX: $int> TryFrom<$int> for $ty<MIN, MAX> {
+            type Error = $err<MIN, MAX>;
+
+            fn try_from(value: $int) -> Result<Self, Self::Error> {
+                if (MIN..MAX).contains(&value) {
+                    Ok(Self(value))
+                } else {
+                    Err($err(value))
+                }
+            }
+        }
+
+        impl<const MIN: $int, const MAX: $int> Packable for $ty<MIN, MAX> {
+            type PackError = Infallible;
+            type UnpackError = $err<MIN, MAX>;
+
+            fn packed_len(&self) -> usize {
+                self.0.packed_len()
+            }
+
+            fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), PackError<Self::PackError, P::Error>> {
+                self.0.pack(packer)
+            }
+
+            fn unpack<U: Unpacker>(unpacker: &mut U) -> Result<Self, UnpackError<Self::UnpackError, U::Error>> {
+                $int::unpack(unpacker)
+                    .infallible()?
+                    .try_into()
+                    .map_err(UnpackError::Packable)
+            }
+        }
+    };
+}
+
+bounded_int!(
+    /// Wrapper type for a `u8`, providing minimum and maximum value bounds.
```

Comment:
you can use `#[doc = concat!("Dis is ", stringify!($int), "but bounded")]` for docstrings to avoid writing them by hand.

Reply:
Nice, I was looking for a way to do that!

Reply:
It only works on the latest rust version so some CI workflows might complain about it.

Reply:
Yeah, it didn't work :trollface: